### ISSUE TITLE
Possibility for iOS/Android-App?

### ISSUE THREAD

**Issue by votexx** [where2help/where2help#182]
Is there a possibility to facilitate w2h also as an iOS/Android-App (if I´m not mistaken, Benjamin mentioned a wrapper of some kind to make that happen).

We might want to use this later on to send push-requests to volunteers on short notice (like we did with sms in the past). If there is an option, what does it take to set it up and keep the apps (iOS, Android) maintained?

thx...Dominik

**Comment by klappradla** [comment]
If the to-use tool for this was called [Cordova](https://cordova.apache.org/) or [Phonegap](http://phonegap.com/), I personally think it's neither worth the effort and time nor the money to put it up in the App- and Playstore: they may be useful for prototyping, but from what I remember, their performance is just so bad compared to the *normal* mobile browser, that people would probably prefer that anyways. Besides that, getting push notifications to work would be enough additional work to not just *"have a wrapper"* round the rails app. Correct me if I'm wrong @groe.

Depending on the state of both, the [Android](https://github.com/where2help/android) and [iOS](https://github.com/where2help/Where2Help-iOS) app (both not up 2 date from what I know), efforts may be better spent there. But @markusrubey and @ramonh should have more insight here.

I'd for starters concentrate on the web app, its usage-patterns and how users approach it, to then figure out what is really needed. The application has email notifications, so there's a starting point.

**Comment by groe** [comment]
I was referring to the existing iOS and android apps (which I haven't seen so far) assuming that they were webview-based wrappers around the web app but apparently, after having a look in the code now, they are actually full-fletched native apps.

I don't know about the state of the apps though (iOS has some recent commits from @ramonh, Android > 6 months ago), so I can't say much about the client-side efforts for push notifications. On the server-side I have used [rpush](https://github.com/rpush/rpush) in other projects, which works good but setting everything up, creating an API for clients to register their device tokens, handle multi-device issues, etc. does involve some effort.

Were the apps also developed during the hackathon or after it? By the same team? Whats the goal with the apps? What (part of the) functionality do or should they offer? Whats their current state? (And who can answer these questions? :grinning:  )

**Comment by klappradla** [comment]
Thanx for pointing out *rpush* @groe, did not know about this, but looks kinda promising ✌️ 

As for the mobile apps: the Android app was part also part of the hackathon, but is developed by just a single person (@markusrubey) and with no updates within the last 6 months definitely not up to date with our API endpoint. The iOS app was started by Aaron in the last hours of the hackathon and recently taken on by @ramonh. It speaks to the current state of our API endpoint, but does not have all required features in yet.
Speaking of features: the mobile apps provide all basic features a **volunteer** needs to use the platform: sign in, browsing events, opting into events, viewing ones schedule. They don't provide an interface for neither admins nor NGOs.

**Comment by votexx** [comment]
> I'd for starters concentrate on the web app, its usage-patterns and how users approach it, to then figure out what is really needed. The application has email notifications, so there's a starting point.

@klappradla: Is there a possibility to trigger the email-notification directly through a job? E.g. when we need volunteers on short notice, that we can send an email to all registered users?

**Comment by klappradla** [comment]
Theoretically yes, @votexx, but it is currently not implemented like this. For doing what you describe, we would need to come up with:
- a setting for users to enable/disable these types of *"push"* notifications
- agree upon how these emails are triggered and send
- agree upon the content, e.g. wether it should be written by hand or auto generated

A scenario and workflow I can imagine would be: having an **admin-user-only** button on events or shifts (yet to decide), something like **`invite all volunteers`** which could trigger an email invitation to the event/shift, automatically using the users' set language.

Would that solve the problem you want to solve?

**Comment by votexx** [comment]
Thx for the suggestion @klappradla - the feature should, if possible, look as follows:
- Checkbox "alle Freiwilligen einladen" that **can** be available for all events
- **can**, because this function should only be available, if an admin activates it (so it can be controlled, if we have emergency situations, or not)
- the terms of use state, that registered users agree to be informed on short notice via email in case of "emergency" - so there is no need for a checkbox in users´ profile

Also the question: if possible, how long does it take? FSW want´s to know, to be prepared for the presentation taking place on 22.9. to inform about "will be available soon"

**Comment by klappradla** [comment]
Implementing something like this would take me about a week, already taken into account that I can only work on it at nights. However, I don't fully get your vision of this feature yet, @votexx.

A *checkbox* is not an action item, selecting one shouldn't trigger something as big and important such as sending email to all users. A classic workflow would be: have a button **`Alle Freiwilligen einladen`** and after clicking it a confirmation dialog if one is really sure.

Please clearify:
> function should only be available, if an admin activates it (so it can be controlled, if we have emergency situations, or not)

Do you mean, admins should on a case by case base be able to *"unlock"* this feature for certain Events in order for NGOs to be able to press the button then themselves?

**Comment by markusrubey** [comment]
**State of the Android App**

As you might have guessed, the Android repository here on Github is not up to date. I was not happy with the results of the hackathon and quite a lot changed since then, so I started from scratch a couple months ago - as the web app team did. I was not active the last two months or so due to work and moving. This means I would need at least another month to finish the app. Further, I first need to get an overview of the current project state before I can say more about timing.

Good news: I'm now able to find time to work on where2help again. Mostly on Fridays or random late evenings.

**Feedback regarding Notifications**

Since both iOS and Android projects are native apps it would make sense to provide platform tailored notifications. Google provides a simple solution called GCM which is basically free and supports both platforms. Again, I don't know timelines so this might not fit our current schedule.

**Comment by votexx** [comment]
> Do you mean, admins should on a case by case base be able to "unlock" this feature for certain Events in order for NGOs to be able to press the button then themselves?

@klappradla It´s not a case by case function, but a global one - so, if activated by an admin, the button you suggested should be available for all events/ngos

**Comment by klappradla** [comment]
Ok @votexx, got you on this! What policy did you think of for activating and deactivating the feature? Provide a global toggle in the admin interface? Or an *application settings* section in the admin interface?

A problem I still see: it's fully up to the NGO to judge an event as urgent or not. I could go ahead, create 10 events starting in one year and within a few clicks spam all users about this.

An alternative and in my eyes more user friendly solution would be: send out (if necessary daily) automated emails to all users, informing them about events which still need people - only if there are any. Plus also provide an option for users to opt out of this feature. Doing so, we would not need to have a human being - the admin - constantly watching the list of events, but still get the desired result: people knowing about which events are current and urgent without deliberately coming to the website at first.

**Comment by votexx** [comment]
Is there any news on the iOS-App (if it´s not in here, maybe there are commits I can´t see)?! 

@klappradla regarding your suggested "mailing-list": like that idea...maybe with an link to unsubscribe from the automated emails. So an admin can switch on the function and the mails start automatically on a regular basis - if not needed anymore, we can switch it off again, right?

**Comment by klappradla** [comment]
The iOS app lives in this repo: https://github.com/where2help/Where2Help-iOS

However, there are no news, development is currently stopped on it afaik.

Regarding the *"mailing-list"*: almost ✌️ 
I would not implement a switch in the admin interface. If the function is no longer needed, one can just change the schedule in the code (depends on what @groe wants to use for it - maybe [whenever](https://github.com/javan/whenever) a timeless classic). As I said, I don't really know how such a *"settings"* interface for admins could look like and don't have an image in my head yet, how I would wire it up to functions like this. Redeploying a new app version is definitely the short hand approach here.

**Comment by ramonh** [comment]
Some news on the iOS app (oh my, it's been a month now since I last touched it 😰)

The app now supports opting in and out of shifts, as well as a fully done login, logout and registration. This also supports using the built in iOS sharing functionality.

What's still missing is an overview on what you're opted to help in, as well as editing your profile. Other than that, I've never really looked into how iOS' notifications work. I think you need a separate server sending those though!

**Comment by votexx** [comment]
So the "mailing-list" can be implemented - it´s just a question of how to switch it on and off.
Thx for getting back to us @ramonh - so if the "my shifts" overview can be implemented, the app provides the same functions as the WebApp?

**Comment by groe** [comment]
@klappradla Yup, `whenever` is great! It would invoke a rake task that does the sending. Would you set up the rake task in a way that keeps track of when the last email was sent or would it *always* send out the mails when the task is invoked?

**Comment by klappradla** [comment]
I don't have a clue yet, how the feature can look like in its implementation @groe, sorry. Would also be open for suggestions on an admin settings panel, e.g. examples how this could look like.

But in general: extracting these things into a rake task will probably be the approach. I'm not sure if we can use something like `whenever` on the free Heroku plan and rake tasks are also compatible with Heroku's scheduler add-on.
